javabuilder-authorizer/lambda_function.rb

```diff
@@ -0,0 +1,78 @@
+require 'aws-sdk-lambda'
+require 'jwt'
+
+# The lambda handler takes an event with the header
+# 'Authorization: Bearer <token>',  where the token is a JWT token
+# generated by dashboard. The handler verifies the token with the appropriate
+# public key (dependant on the environment the request came from),
+# and checks the token has not expired and its issue time is not in the future.
+def lambda_handler(event:, context:)
+  headers = event['headers']
+  auth_header = headers['Authorization']
+  method_arn = event['methodArn']
+
+  if auth_header.start_with?('Bearer: ')
+    jwt_token = auth_header.split[1]
+    begin
+      decoded_token = JWT.decode(
+        jwt_token,
+        get_public_key(method_arn),
+        true,
+        verify_iat: true, # verify issued at time is valid
+        algorithm: 'RS256'
+      )
+    rescue JWT::ExpiredSignature, JWT::InvalidIatError
+      return generate_deny(nil, method_arn)
+    end
+
+    return generate_deny(nil, method_arn) unless decoded_token
+
+    generate_allow('me', method_arn, decoded_token[0])
+  else
+    generate_deny(nil, method_arn)
+  end
+end
+
+# Helper function to generate an IAM policy.
+def generate_policy(principal_id, effect, resource, token_data)
+  # Required output:
+  auth_response = {}
+  auth_response['principalId'] = principal_id
+  if effect && resource
+    policy_document = {}
+    policy_document['Version'] = '2012-10-17' # default version
+    policy_document['Statement'] = []
+    statement_one = {}
+    statement_one['Action'] = 'execute-api:Invoke' # default action
+    statement_one['Effect'] = effect
+    statement_one['Resource'] = resource
+    policy_document['Statement'][0] = statement_one
+    auth_response['policyDocument'] = policy_document
+  end
+  auth_response['context'] = token_data if token_data
+  auth_response
+end
+
+def generate_allow(principal_id, resource, token_data)
+  generate_policy(principal_id, 'Allow', resource, token_data)
+end
+
+def generate_deny(principal_id, resource)
+  generate_policy(principal_id, 'Deny', resource, nil)
+end
+
+# Load the JWT token from public_keys based on the environment the request
+# was sent from.
+#
+# method_arn is in format
+# arn:aws:execute-api:region:account-id:api-id/stage-name/$connect
+# We only care about stage--that tells us the environment (development, staging, etc.)
+def get_public_key(method_arn)
+  tmp = method_arn.split(':')
+  api_gateway_arn = tmp[5].split('/')
+  stage_id = api_gateway_arn[1]
```

Comment:
Minor NIT: `stage_id` is used here, but `stage-name` is used in the comment. It took me a bit to realize they were the same thing.

Reply:
unified around stage_name, which makes more sense.